### ISSUE TITLE
org dates are not correctly exported to dates in odt

### ISSUE THREAD

**Issue by ouboub** [kjambunathan/org-mode-ox-odt#244]
Hi

while I have no problem exporting org-tables with org-dates to ods, when I try to export
an org document with org-dates and dates in org-tables, they are not exported correctly.
I attach an org file, with the exported odt, and ods converted versions.
I am sure I miss something, but can't remember what.
regards
Uwe Brauer 
[Date-Prob.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/12168707/Date-Prob.zip)

**Comment by emacksnotes** [comment]
I have pushed a fix. 

I remember we having some discussion about locale-specific rendering of date in gnumeric / libreoffice calc.

As this bug illustrates, the formatting of date fields affects both ODT and ODS export.

If you want locale-specific rendering in ODT export then flip  `number:automatic-order` to `false` here

https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/lisp/ox-odt.el#L3546

There may be other places where the date style might be using automatic ordering of date fields.  (Automatic ordering means, the ordering is affected by the LOCALE of your LIBREOFFICE INSTALLATION and not the LOCALE of EMACS or its customizations)

Let me know what you find.  Keep the bug open, I will audit date related styles at my own face.  But if you point out specific issues with example, I can fix them right away.

**Comment by kjambunathan** [comment]
I have pushed a fix. 

I remember we having some discussion about locale-specific rendering of date in gnumeric / libreoffice calc.

As this bug illustrates, the formatting of date fields affects both ODT and ODS export.

If you want locale-specific rendering in ODT export then flip  `number:automatic-order` to `false` here

https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/lisp/ox-odt.el#L3546

There may be other places where the date style might be using automatic ordering of date fields.  (Automatic ordering means, the ordering is affected by the LOCALE of your LIBREOFFICE INSTALLATION and not the LOCALE of EMACS or its customizations)

Let me know what you find.  Keep the bug open, I will audit date related styles at my own face.  But if you point out specific issues with example, I can fix them right away.

**Comment by kjambunathan** [comment]
> I always obtain the format 26/07/2023 while I would prefer 26.07.2023 as in the ods export.

You aren't saying if you flipped the following boolean _locally_ or not ...

> In this line https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/lisp/ox-odt.el#L3546 change `true` to `false`.

Once you flip that boolean locally, your ODT is fine, but ODS has issues, right?

(I have trouble grasping what you are saying, and you may have to re-state what you are saying)

**Comment by kjambunathan** [comment]
Try flipping the value of `automatic-order` in the following places

https://github.com/search?q=repo%3Akjambunathan%2Forg-mode-ox-odt+number%3Aautomatic-order+path%3A%2F%5Eetc%5C%2Fstyles%5C%2F%2F&type=code

**Comment by kjambunathan** [comment]
This is how you identify the styles that would be of interest to you:

In the `content.xml` of `ods` file search for date of interest, say `C-s 2023`. 

```xml
	    <table:table-cell office:date-value="2023-07-26T00:00:00"
			      office:value-type="date"
			      table:style-name="OrgTableCellTDate">
	      <text:p text:style-name="OrgTableContentsLeft"></text:p>
	    </table:table-cell>
```

The style you see is `OrgTableCellTDate`

If you search for the above style in `content.xml` AND `styles.xml` you will see

```xml
    <style:style style:name="OrgTableCellTDate"
		 style:parent-style-name="OrgTableCellT"
		 style:family="table-cell"
		 style:data-style-name="OrgDateStyle" />

```

You can see two styles `OrgTableCellT` and `OrgDateStyle`.

 `OrgTableCellT`  is like this

```xml

    <style:style style:name="OrgTableCellT"
		 style:family="table-cell">
      <style:table-cell-properties fo:border-bottom="none"
				   fo:border-left="none"
				   fo:border-right="none"
				   fo:border-top="0.002cm solid #000000"
				   fo:padding="0.159cm" />
    </style:style>

```

and 

`OrgDateStyle` is like this

```xml
    <number:date-style style:name="OrgDateStyle"
                       number:automatic-order="true">
      <number:day number:style="long" />
      <number:text>/</number:text>
      <number:month number:style="long" />
      <number:text>/</number:text>
      <number:year />
    </number:date-style>
```

and defined in `styles.xml`.

... and it has `automatic-order` ON.

And `OrgDateStyle` comes from the factory styles which is located in one of the `xml` files under `styles`  

```
~/src1/org-mode-ox-odt/master/etc/styles$  tree .
.
├── ods
│   ├── content.xml
│   ├── META-INF
│   │   └── manifest.xml
│   ├── meta.xml
│   ├── mimetype
│   ├── OrgOdsTemplate.ots
│   └── styles.xml
├── OrgOdtContentTemplate.xml
├── OrgOdtStyles.xml
└── README

3 directories, 9 files
```

The root directory has `odt` styles, and the `ods` subdir has `ods` styles.

**Comment by kjambunathan** [comment]
Based on the above notes, this is a possible fix for your problem  ...

In https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/etc/styles/ods/styles.xml#L75

change 

```xml
  <number:date-style style:name="OrgDateStyle" number:automatic-order="true">
    <number:day number:style="long" />
    <number:text>/</number:text>
    <number:month number:style="long" />
    <number:text>/</number:text>
    <number:year />
  </number:date-style>

```

to 

```xml

  <number:date-style style:name="OrgDateStyle"
		     >
    <number:month number:style="long" />
    <number:text>.</number:text>
    <number:day number:style="long" />
    <number:text>.</number:text>
    <number:year />
    </number:date-style>
```

In the re-written styles, I have


- turned OFF `automatic-order`
- Change the field separator from `/` to `.`
- re-arranged the date fields so that month comes first etc.

Experiment with this fix "by hand",  and if it works for you

- I will add support for custom `in buffer` `nxml` styles for `ods` export
- Make sure that `OrgDateStyle` is based on custom fields

Waiting for your feedback ...

**Comment by ouboub** [comment]
> > I always obtain the format 26/07/2023 while I would prefer 26.07.2023 as in the ods export.
> 
> You aren't saying if you flipped the following boolean _locally_ or not ...
> 
> > In this line https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/lisp/ox-odt.el#L3546
> >  change `true` to `false`.
> 
> Once you flip that boolean locally, your ODT is fine, but ODS has issues, right?
> 
> (I have trouble grasping what you are saying, and you may have to re-state what you are saying)

Sorry I wrote that to hastly.
I set the value to false, that is

```
	      (concat " number:automatic-order=\"false\""
```
and recompiled

but the date format in the odt document still was not changed, but remained 26/07/2023
the exported ods file is ok.
I just see that you sent a couple of other posts so I will try them out.

**Comment by ouboub** [comment]
> Based on the above notes, this is a possible fix for your problem ...
> 
> In
> 
> https://github.com/kjambunathan/org-mode-ox-odt/blob/433c3c938f3327b778a2a1d4b18aab6d011f5d8e/etc/styles/ods/styles.xml#L75
> 
> change
> 
> ```
>   <number:date-style style:name="OrgDateStyle" number:automatic-order="true">
>     <number:day number:style="long" />
>     <number:text>/</number:text>
>     <number:month number:style="long" />
>     <number:text>/</number:text>
>     <number:year />
>   </number:date-style>
> ```
> 
> to
> 
> ```
>   <number:date-style style:name="OrgDateStyle"
> 		     >
>     <number:month number:style="long" />
>     <number:text>.</number:text>
>     <number:day number:style="long" />
>     <number:text>.</number:text>
>     <number:year />
>     </number:date-style>
> ```
> 
> In the re-written styles, I have
> 
> * turned OFF `automatic-order`
> * Change the field separator from `/` to `.`
> * re-arranged the date fields so that month comes first etc.
> 
> Experiment with this fix "by hand", and if it works for you
> 
> * I will add support for custom `in buffer` `nxml` styles for `ods` export
> * Make sure that `OrgDateStyle` is based on custom fields
> 
> Waiting for your feedback ...

I did the change you recommend, but it seemed not to have helped, attached you find the results

I will now try out your last recommendation
[Date-Prob.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/12174843/Date-Prob.zip)

**Comment by kjambunathan** [comment]
For `ods` what matters is the value of `OrgDateStyle`.

For `odt` what matters are the values of `OrgDate1` and `OrgDate2`.

The XML files generated by the exporter is one long line.  I am surprised how you could get a pretty printed xml so fast.    Surely you are using a tool other than Emacs to view XML files.  What is it?

**Comment by ouboub** [comment]
> For `ods` what matters is the value of `OrgDateStyle`.
> 
> For `odt` what matters are the values of `OrgDate1` and `OrgDate2`.
> 
> The XML files generated by the exporter is one long line. I am surprised how you could get a pretty printed xml so fast. Surely you are using a tool other than Emacs to view XML files. What is it?

```
xmllint --format content.xml > content2.xml
```

**Comment by kjambunathan** [comment]
A quck tip ...

ODT exporter uses tidy for pretty printing.

Use M-x apropos-option  odt prett

to identify the variable

Or

C-c C-e # o

to change the per buffer option

(Pretty printing is only for debugging.)

I will make some elisp changes su that debugging and experimentation
becomes easy ...

I will message once that is done

**Comment by ouboub** [comment]
I run another check.
in bash I run 
LANG=de_DE.UTF-8 LANGUAGE=de_DE.UTF-8
I started swriter (7,2)
indeed german is the default language as the attached screenshots shows.
But opening the converted odt the dates are displayed by english/american standard, so switching back to german 
solves the issue.
BTW, any chance to get rid of the < > as in the ods exporter?

![swriter](https://github.com/kjambunathan/org-mode-ox-odt/assets/5957203/1315e27b-4825-4eaf-a2ac-7e89510e415e)

![swriter1](https://github.com/kjambunathan/org-mode-ox-odt/assets/5957203/dfd7885a-d2a1-4894-9863-9b998afde00e)

**Comment by ouboub** [comment]
>  BTW, any chance to get rid of the < > as in the ods exporter?

Sorry my bad, the < > are deleted when exported, I just did not do it properly, the English date problem still persists, though

**Comment by kjambunathan** [comment]
A note to future me ...

Date fields have been in ODT exporter forever.  I think I implemented with the assumption that one can click on it and subsequently modify it with a Libreoffice calendar widget.  

The recent experimentations shows that with both and ODT and ODS, having a calendar widget pop up is not a default action. It requiers further configuration or installing additional extensions.

I strongly remember that if one is designing `forms` (human fillable `forms` that you submit to the bureaucratic machinery) it is possible to have calendar control pop up on clicking the date field.

In other words, the only advantage of date field that I see is math on dates with Calc.  And if your or my use-case doesn't involve much math with date or date time objects, then date as a string works admirably with less head-scratching.

What I am saying is one can get by without date fields, ... but I want to have  date fields around (and experiment with it) with the hope that exporter will continue to grow in useful ways.

**Comment by kjambunathan** [comment]
[issue-244-timestamps-unit-test.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/12267595/issue-244-timestamps-unit-test.zip)

---------

To get this

![Screenshot from 2023-08-05 20-27-06](https://github.com/kjambunathan/org-mode-ox-odt/assets/19869864/28eaf3da-22b3-45fe-b503-b293c9fe9c48)

do this

```
#+odt_preferred_output_format: pdf

# #+ODT_PRETTIFY_XML: tidy+indent

#+ATTR_ODT: :style "GriddedTable"
#+ATTR_ODT: :widths "1,1"
#+CAPTION: simple
| <2023-08-02 Wed> | <2023-08-02 Tue 20:00-22:00> |

#+name: customize
#+begin_src emacs-lisp :results silent :exports results
  (require 'ox-ods)
  (custom-set-variables
   '(org-display-custom-times t)
   '(org-time-stamp-custom-formats '("%d-##-%m-##-%Y" . "%d-##-%m-##-%Y %I-**-%M"))
   ;; Dates as text
   '(org-odt-timestamp-options '( :strip-brackets-around-date t
				  :emit-date-as-date-field nil
				  :emit-date-as-date-object nil
				  :number:date-style-attributes
				  ( :number:automatic-order nil)))
   )
#+end_src

#+name: settings
#+begin_src emacs-lisp :exports results
  `(
    ,@`(
	(org-time-stamp-formats ,org-time-stamp-formats)
	(org-display-custom-times ,org-display-custom-times)
	(org-time-stamp-custom-formats ,org-time-stamp-custom-formats)
	)
    ,@'((org-odt-timestamp-options))
    ,@(seq-map
       (lambda (it)
	 (list it (plist-get org-odt-timestamp-options it))
	 )
       '( :strip-brackets-around-date
	  :emit-date-as-date-field
	  :emit-date-as-date-object
	  :number:date-style-attributes)))
#+end_src

#+caption: Settings
#+ATTR_ODT: :style "GriddedTable"
#+ATTR_ODT: :widths "8,10"
#+RESULTS: settings
| org-time-stamp-formats        | (%Y-%m-%d %a . %Y-%m-%d %a %H:%M)          |
| org-display-custom-times      | t                                          |
| org-time-stamp-custom-formats | (%d-##-%m-##-%Y . %d-##-%m-##-%Y %I-**-%M) |
| org-odt-timestamp-options     |                                            |
| :strip-brackets-around-date   | t                                          |
| :emit-date-as-date-field      | nil                                        |
| :emit-date-as-date-object     | nil                                        |
| :number:date-style-attributes | (:number:automatic-order nil)              |

* COMMENT Complex Examples <2023-08-01 Tue>

#+ATTR_ODT: :style "GriddedTable"
#+ATTR_ODT: :rel-width 100
#+CAPTION: Table with Timestamps
| <2006-11-01 Wed>                   | [2006-11-01 Wed]                   |
| <2006-11-01 Wed 19:15>             | [2006-11-01 Wed 19:15]             |
| <2006-11-02 Thu 20:00-22:00>       | [2006-11-02 Thu 20:00-22:00]       |
| <2007-05-16 Wed 12:30 +1w>         | [2007-05-16 Wed 12:30 +1w]         |
| <%%(diary-float t 4 2)>            | [%%(diary-float t 4 2)]            |
| <2004-08-23 Mon>--<2004-08-26 Thu> | [2004-08-23 Mon]--[2004-08-26 Thu] |
| <2005-10-01 Sat +1m>               | [2005-10-01 Sat +1m]               |
| <2005-10-01 Sat +1m -3d>           | [2005-10-01 Sat +1m -3d]           |
| <2008-02-10 Sun ++1w>              | [2008-02-10 Sun ++1w]              |
| <2008-02-08 Fri 20:00 ++1d>        | [2008-02-08 Fri 20:00 ++1d]        |
| <2005-11-01 Tue .+1m>              | [2005-11-01 Tue .+1m]              |
| <2019-04-05 08:00 Fri .+1h>        | [2019-04-05 08:00 Fri .+1h]        |

** Plain timestamp;
  <2006-11-01 Wed 19:15>

** Plain timestamp; Event; Appointment
  <2006-11-02 Thu 20:00-22:00>

** Timestamp with repeater interval
  <2007-05-16 Wed 12:30 +1w>

** 22:00-23:00 Diary-style expression entries
  <%%(diary-float t 4 2)>

** Time/Date range
   <2004-08-23 Mon>--<2004-08-26 Thu>

** Inactive timestamp
  [2006-11-01 Wed]

** TODO Repeater
   DEADLINE: <2005-10-01 Sat +1m>

** TODO Repeater and Warning
   DEADLINE: <2005-10-01 Sat +1m -3d>

** TODO Special Repeater 1
   DEADLINE: <2008-02-10 Sun ++1w>

   Marking this DONE shifts the date by at least one week, but also
   by as many weeks as it takes to get this date into the future.
   However, it stays on a Sunday, even if you called and marked it
   done on Saturday.

** TODO Special Repeater 2
   DEADLINE: <2008-02-08 Fri 20:00 ++1d>

Marking this DONE shifts the date by at least one day, and also
   by as many days as it takes to get the timestamp into the future.
   Since there is a time in the timestamp, the next deadline in the
   future will be on today's date if you complete the task before
   20:00.

** TODO Special Repeater 3
   DEADLINE: <2005-11-01 Tue .+1m>

   Marking this DONE shifts the date to one month after today.

** TODO Special Repeater 4
   DEADLINE: <2019-04-05 08:00 Fri .+1h>

   Marking this DONE shifts the date to exactly one hour from now.

#+odt_extra_styles: <style:style style:name="OrgActiveTimestamp"
#+odt_extra_styles:              style:parent-style-name="OrgTimestamp"
#+odt_extra_styles:              style:family="text">
#+odt_extra_styles:   <style:text-properties fo:background-color="transparent"
#+odt_extra_styles: 			 fo:color="#ff0000" />
#+odt_extra_styles: </style:style>
#+odt_extra_styles: <style:style style:name="OrgInactiveTimestamp"
#+odt_extra_styles:              style:parent-style-name="OrgTimestamp"
#+odt_extra_styles:              style:family="text">
#+odt_extra_styles:   <style:text-properties fo:background-color="transparent"
#+odt_extra_styles: 			 style:use-window-font-color="true" />
#+odt_extra_styles: </style:style>
```

**Comment by kjambunathan** [comment]
Leave the issue open. 

 - I  have to update the docstring
 - amend the commit message.  
 -  report a bug in libreoffice calc

**Comment by kjambunathan** [comment]
May be clocktables are good candidates for exporting to ODS.  
 See [Org Mode: Exporting Clock Tables](https://olddeuteronomy.github.io/post/org-export-clocktables/)

Specifically this line 

> The problem is that my employer wants the report to be in a specific format.

This essentially means using custom display of date and date time.

It is good to mentally run through the above case, and see if certain `ox-ods` functionalities could be tested against or improved.

**Comment by ouboub** [comment]
>>> "JK" == Jambunathan K ***@***.***> writes:

> [issue-244-timestamps-unit-test.zip](https://github.com/kjambunathan/org-mode-ox-odt/files/12267595/issue-244-timestamps-unit-test.zip)
> ---------

> To get this

> ![Screenshot from 2023-08-05 20-27-06](https://github.com/kjambunathan/org-mode-ox-odt/assets/19869864/28eaf3da-22b3-45fe-b503-b293c9fe9c48)

> do this

I pull the latest commits, 
run 

Make clean
make

Restarted emacs

Unzipped the attached zip.

Exported to odt, was puzzled at first, changed then the setting back to my preferred

 '(org-time-stamp-custom-formats '("%d.%m.%Y" . "%d.%m.%Y %I:%M"))

Exported again, and the result looks as I expected, but the date no
longer is a date field,  which is fine for me


I played around a bit with 


:emit-date-as-date-object t ; not sure that does 


Not sure that the last variable is supposed to do. 
In any case, thanks for that solution. It will be helpful in my
workflow.

regards

**Comment by kjambunathan** [comment]
> Exported to odt, was puzzled at first, changed then the setting back to my preferred

>  '(org-time-stamp-custom-formats '("%d.%m.%Y" . "%d.%m.%Y %I:%M"))


It is just a format string, you can put whatever you want.  I added an exotic format with fancy separator to check if the `OrgDate` and `OrgDateAndTime` styles are kicking in.

In the `org` snippet, the date is shared textually ... so the date styles have no bearing.

> and the result looks as I expected, but the date no longer is a date field,  which is fine for me

If someone needs date arithmetic on spreadsheet it would come in handy.   

Many of these knobs are for me to experiment with.
